otherlibs/systhreads/decs.c

```diff
@@ -0,0 +1,801 @@
+struct caml_dec_struct {
+
+  value descr;
+
+  struct caml_dec_struct * next;
+  struct caml_dec_struct * prev;
+
+  int domain_id;
+
+  struct stack_info* current_stack;
+  struct c_stack_link* c_stack;
+  struct caml__roots_block *local_roots;
+  struct longjmp_buffer *exit_buf;
+  int backtrace_pos;
+  code_t * backtrace_buffer;
+  caml_root backtrace_last_exn;
+  value * gc_regs_buckets;
+  value ** gc_regs_slot;
+  void * exn_handler;
+
+  #ifndef NATIVE_CODE
+  intnat trap_sp_off;
+  intnat trap_barrier_off;
+  struct caml_exception_context* external_raise;
+  #endif
+
+};
+
+typedef struct caml_dec_struct* caml_dec_t;
+
+/* overall table for decs accross domains */
+struct caml_dec_table {
+  caml_dec_t all_decs;
+  caml_dec_t current_dec;
+  pthread_key_t dec_key;
+  struct caml_dec_main_lock dec_lock;
+};
+
+/* dec_table instance, up to Max_domains */
+static struct caml_dec_table dec_table[Max_domains];
+
+/* unique dec id */
+static atomic_uintnat dec_next_id;
+
+#define Dec_main_lock dec_table[Caml_state->id].dec_lock
+#define Dec_key dec_table[Caml_state->id].dec_key
+#define All_decs dec_table[Caml_state->id].all_decs
+#define Current_dec dec_table[Caml_state->id].current_dec
+
+static void (*prev_scan_roots_hook) (scanning_action, void *, struct domain *);
+
+/* main lock's logic, analogous to systhread's master lock. */
+/* The main lock is to be held by a dec that is going to execute OCaml code */
+/* or interfere with the runtime in any way. */
+/* There is one main lock per domain. */
+
+static void caml_dec_main_lock_init(struct caml_dec_main_lock *m) {
+
+  caml_plat_mutex_init(&m->lock);
+  caml_plat_cond_init(&m->free, &m->lock);
+  atomic_store_rel(&m->busy, 1);
+  atomic_store_rel(&m->waiters, 0);
+  return;
+};
+
+static void caml_dec_main_lock_acquire(struct caml_dec_main_lock *m)
+{
+  caml_plat_lock(&m->lock);
+  while (atomic_load_acq(&m->busy)) {
+    atomic_fetch_add(&m->waiters, +1);
+    caml_plat_wait(&m->free);
+    atomic_fetch_add(&m->waiters, -1);
+  }
+  atomic_store_rel(&m->busy, 1);
+  caml_plat_unlock(&m->lock);
+}
+
+static void caml_dec_main_lock_release(struct caml_dec_main_lock * m)
+{
+  caml_plat_lock(&m->lock);
+  atomic_store_rel(&m->busy, 0);
+  caml_plat_signal(&m->free);
+  caml_plat_unlock(&m->lock);
+}
+
+static void caml_dec_main_lock_yield(struct caml_dec_main_lock * m)
+{
+  uintnat waiters;
+
+  caml_plat_lock(&m->lock);
+  /* We must hold the lock to call this. */
+
+  /* We already checked this without the lock, but we might have raced--if
+     there's no waiter, there's nothing to do and no one to wake us if we did
+     wait, so just keep going. */
+  waiters = atomic_load_acq(&m->waiters);
+
+  if (waiters == 0) {
+    caml_plat_unlock(&m->lock);
+    return;
+  }
+
+  atomic_store_rel(&m->busy, 0);
+  caml_plat_signal(&m->free);
+  atomic_fetch_add(&m->waiters, +1);
+  do {
+    /* Note: the POSIX spec prevents the above signal from pairing with this
```

Comment:
Just so I understand this correctly, what part of the POSIX spec prevents the above signal from hitting the wait below?

Reply:
This part of the logic is imported from `systhreads`, especially this implementation comes from https://github.com/ocaml/ocaml/pull/2112
When I reworked this code for multicore I looked into the specifics, and I think it means that a `signal` will only signal to currently `wait`ing threads. (if no waiters are available, it will effectively do nothing, noting that there's also a specific handling for this at the start of the function for what I gathered was optimization purpose).
As such there's not really a way this `signal` can pair with the following `wait`.
But maybe I missed something more subtle.

Slightly related, but since the linked PR does mention this, I think some locking logic may benefit from a rework since the previous use of `volatile` is gone with the use of C11 atomic operations through `platform.h`.